src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Setup/SetupComponentRegistrationService.cs

```diff
@@ -0,0 +1,259 @@
+    private readonly IVsService<SVsSetupCompositionService, IVsSetupCompositionService> _setupCompositionServiceFactory;
+    private readonly ISolutionService _solutionService;
+    private readonly IProjectFaultHandlerService _projectFaultHandlerService;
+    private readonly Lazy<HashSet<string>?> _installedRuntimeComponentIds;
+
+    private IVsSetupCompositionService? _setupCompositionService;
+    private IMissingComponentRegistrationService? _missingComponentRegistrationService;
+    private IAsyncDisposable? _solutionEventsSubscription;
+    private ImmutableDictionary<Guid, UnconfiguredSetupComponentSnapshot?> _snapshotByProjectGuid = ImmutableDictionary<Guid, UnconfiguredSetupComponentSnapshot?>.Empty;
+
+    [ImportingConstructor]
+    public SetupComponentRegistrationService(
+        IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> serviceBrokerContainer,
+        IVsService<SVsSetupCompositionService, IVsSetupCompositionService> vsSetupCompositionService,
+        ISolutionService solutionService,
+        IProjectFaultHandlerService projectFaultHandlerService,
+        JoinableTaskContext joinableTaskContext)
+        : base(new(joinableTaskContext))
+    {
+        _serviceBrokerContainer = serviceBrokerContainer;
+        _setupCompositionServiceFactory = vsSetupCompositionService;
+        _solutionService = solutionService;
+        _projectFaultHandlerService = projectFaultHandlerService;
+
+        _installedRuntimeComponentIds = new Lazy<HashSet<string>?>(FindInstalledRuntimeComponentIds);
+
+        static HashSet<string>? FindInstalledRuntimeComponentIds()
+        {
+            // Workaround for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1460328
+            // VS Setup doesn't know about runtimes installed outside of VS. Deep detection is not suggested for performance reasons.
+            // Instead we read installed runtimes from the registry.
+
+            // TODO Consider the architecture of the project itself.
+            // We currently assume that the project will run on the same architecture as VS.
+            // This will be the common case, but does not cover situations where a project runs
+            // under emulation (e.g. an x64 build running on an ARM64 system).
+            Architecture architecture = RuntimeInformation.ProcessArchitecture;
+
+            string[]? runtimeVersions = NetCoreRuntimeVersionsRegistryReader.ReadRuntimeVersionsInstalledInLocalMachine(architecture);
+
+            if (runtimeVersions is null)
+            {
+                return null;
+            }
+
+            HashSet<string> installedRuntimeComponentIds = new(StringComparers.VisualStudioSetupComponentIds);
+
+            foreach (string runtimeVersion in runtimeVersions)
+            {
+                // Example versions:
+                //
+                // - "3.1.32"
+                // - "7.0.11"
+                // - "8.0.0-preview.7.23375.6"
```

Comment:
Are there any other version suffixes other than preview or rc?

Reply:
I don't know. There may have been others in the past, and there may be others in the future. The key is that there can be `.` characters in the suffix, and the previous `LastIndexOf('.')` code wouldn't have worked. I like to include examples so that bugs like that are easier to spot. You can check the registry on your machine for other kinds of examples to include here.